### PR TITLE
Fix ramp animated values

### PR DESCRIPTION
**Pull request by elenatorro**
Related with https://github.com/CartoDB/carto-vl/issues/754

Fixed:

```js
color: ramp(linear(log($numfloors), 2, 4), [red, rgb(0,0,255 * sin(now()))])
```

ramp_animated

**Progress**

* [x] Fix children chain for ramps
* [x] Fix animation working properly in children chain
* [x] Add specs

### CODE REVIEW

File: src/renderer/viz/expressions/ramp.js [hunk 1]
```diff
@@ -287,28 +284,31 @@ export default class Ramp extends BaseExpression {
     _computeGLTextureIfNeeded (gl) {
         const texturePixels = this._computeTextureIfNeeded();
 
-        if (this._GLtexCategories !== this.input.numCategories) {
+        if (this._GLtexCategories !== this.input.numCategories || this.isAnimated()) {
```
Comment (davidmanzanares) [inline src/renderer/viz/expressions/ramp.js]: Uploading the texture is performance intensive and I think this is only needed (regarding animation) when the palette is animated, not when the rest of the ramp parameters are animated.

File: src/renderer/viz/expressions/ramp.js [hunk 2]
```diff
@@ -287,28 +284,31 @@ export default class Ramp extends BaseExpression {
...
+            this._bindGLTexture(gl, texturePixels);
         }
     }
 
+    _bindGLTexture (gl, texturePixels) {
+        gl.bindTexture(gl.TEXTURE_2D, this.texture);
+
+        if (this.type === rampTypes.COLOR) {
+            gl.pixelStorei(gl.UNPACK_PREMULTIPLY_ALPHA_WEBGL, false);
+            gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, COLOR_ARRAY_LENGTH, 1, 0, gl.RGBA, gl.UNSIGNED_BYTE, texturePixels);
```
Comment (davidmanzanares) [inline src/renderer/viz/expressions/ramp.js]: This comment may be a premature optimization, but `gl.texImage2D` allocates GPU memory, there is a `gl.texSubImage2D` method that is meant to be used for already allocated texture (see https://www.khronos.org/opengl/wiki/GLAPI/glTexSubImage2D ) The workflow would be to first call `texImage2D` and then, on texture updates, use the `sub` function.

However, if you didn't see a performance problem here, feel free to ignore this.

File: src/renderer/viz/expressions/now.js [hunk 3]
```diff
@@ -28,12 +28,29 @@ export default class Now extends BaseExpression {
...
         super.inlineMaker = inline => inline.now;
     }
-    eval () {
+
+    get value () {
         return this.now.expr;
     }
+
+    eval () {
+        return this.value;
```
[the remark this answers is not in the document]
Comment (elenatorro) [reply src/renderer/viz/expressions/now.js]: I really like this idea and I changed it :) However, I think we should also change then these methods in viewport expressions to be consistent, for example:

```js 
export default class ViewportAvg extends ViewportAggregation {
    constructor (property) {
        super({ property });
        this._sum = 0;
        this._count = 0;
    }

    get value () {
        return this._sum / this._count;
    }

    eval () {
        return this.value;
    }
```

But I am not sure if we should do this in this PR, what do you think?

File: src/renderer/viz/expressions/ramp.js (hunk 2 above)
Comment (elenatorro) [reply src/renderer/viz/expressions/ramp.js]: I do not see a performance problem, but I think this kind of improvements are really great, let me give it a try.

File: src/renderer/viz/expressions/now.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidmanzanares) [reply src/renderer/viz/expressions/now.js]: Sure! that could be in another PR, it's out of the scope, and it's not a priority

**Pull request merged**